File: src/qwenpaw/app/routers/components.py
# -*- coding: utf-8 -*-
"""Explicit/manual component update endpoints."""

from __future__ import annotations

import asyncio

from fastapi import APIRouter, HTTPException

from ...components.service import (
    configured_service,
    queue_all_component_updates,
    queue_component_update,
)
from ...components.update import ComponentUpdateError

router = APIRouter(prefix="/components", tags=["components"])


@router.get("/updates")
async def check_component_updates():
    service = configured_service()
    if service is None:
        return {"enabled": False, "updates": []}
    try:
        return {
            "enabled": True,
            "updates": await asyncio.to_thread(service.check),
        }
    except Exception as exc:  # noqa: BLE001
        raise HTTPException(
            status_code=502,
            detail=f"Component update check failed: {exc}",
        ) from exc
    finally:
        service.client.close()


@router.post("/updates/install")
async def install_all_component_updates():
    """Queue all available managed updates for the next safe startup."""
    try:
        return await asyncio.to_thread(queue_all_component_updates)
    except ComponentUpdateError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    except Exception as exc:  # noqa: BLE001
        raise HTTPException(
            status_code=502,
            detail=f"Component update queue failed: {exc}",
        ) from exc


@router.post("/{component}/install")
async def install_component_update(component: str):
    # Component data may be actively mutated by a loaded plugin.  Until the
    # runtime lifecycle can unload/reload a plugin around the whole snapshot,
    # only the pre-PluginLoader startup path may replace component files.
    try:
        return await asyncio.to_thread(queue_component_update, component)
    except ComponentUpdateError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    except Exception as exc:  # noqa: BLE001
        raise HTTPException(
            status_code=502,
            detail=f"Component update queue failed: {exc}",
        ) from exc

File: console/src/api/modules/components.ts
import { request } from "../request";

export interface ComponentUpdateItem {
  component: string;
  from_version?: string | null;
  target_version?: string;
}

export interface ComponentUpdatesResponse {
  enabled: boolean;
  updates: ComponentUpdateItem[];
}

export interface QueueComponentUpdatesResponse {
  enabled: boolean;
  queued: string[];
  restart_required: boolean;
}

export interface QueueComponentUpdateResponse {
  component: string;
  queued: boolean;
  restart_required?: boolean;
  reason?: string;
}

export const componentsApi = {
  checkComponentUpdates: () =>
    request<ComponentUpdatesResponse>("/components/updates"),
  queueAllComponentUpdates: () =>
    request<QueueComponentUpdatesResponse>("/components/updates/install", {
      method: "POST",
    }),
  queueComponentUpdate: (component: string) =>
    request<QueueComponentUpdateResponse>(
      `/components/${encodeURIComponent(component)}/install`,
      { method: "POST" },
    ),
};
